src/contracts/middleware/BLSRegistryCoordinatorWithIndices.sol

```diff
@@ -0,0 +1,325 @@
+    function getCurrentQuorumBitmapByOperatorId(bytes32 operatorId) external view returns (uint192) {
+        return _operatorIdToQuorumBitmapHistory[operatorId][_operatorIdToQuorumBitmapHistory[operatorId].length - 1].quorumBitmap;
+    }
+
+    /// @notice Returns the number of registries
+    function numRegistries() external view returns (uint256) {
+        return registries.length;
+    }
+
+    /**
+     * @notice Sets parameters of the operator set for the given `quorumNumber`
+     * @param quorumNumber is the quorum number to set the maximum number of operators for
+     * @param operatorSetParam is the parameters of the operator set for the `quorumNumber`
+     */
+    function setOperatorSetParams(uint8 quorumNumber, OperatorSetParam memory operatorSetParam) external onlyServiceManagerOwner {
+        quorumOperatorSetParams[quorumNumber] = operatorSetParam;
+    }
+
+    /**
+     * @notice Registers msg.sender as an operator with the middleware
+     * @param quorumNumbers are the bytes representing the quorum numbers that the operator is registering for
+     * @param registrationData is the data that is decoded to get the operator's registration information
+     * @dev `registrationData` should be a G1 point representing the operator's BLS public key
+     */
+    function registerOperatorWithCoordinator(bytes calldata quorumNumbers, bytes calldata registrationData) external {
+        // get the operator's BLS public key
+        BN254.G1Point memory pubkey = abi.decode(registrationData, (BN254.G1Point));
+        // call internal function to register the operator
+        _registerOperatorWithCoordinator(msg.sender, quorumNumbers, pubkey);
+    }
+
+    /**
+     * @notice Registers msg.sender as an operator with the middleware
+     * @param quorumNumbers are the bytes representing the quorum numbers that the operator is registering for
+     * @param pubkey is the BLS public key of the operator
+     */
+    function registerOperatorWithCoordinator(bytes calldata quorumNumbers, BN254.G1Point memory pubkey) external {
+        _registerOperatorWithCoordinator(msg.sender, quorumNumbers, pubkey);
+    }
+
+    /**
+     * @notice Registers msg.sender as an operator with the middleware
+     * @param quorumNumbers are the bytes representing the quorum numbers that the operator is registering for
+     * @param pubkey is the BLS public key of the operator
+     */
+    function registerOperatorWithCoordinator(
+        bytes calldata quorumNumbers, 
+        BN254.G1Point memory pubkey,
+        OperatorKickParam[] calldata operatorKickParams
+    ) external {
+        require(quorumNumbers.length == operatorKickParams.length, "BLSIndexRegistryCoordinator.registerOperatorWithCoordinator: quorumNumbers and operatorKickParams must be the same length");
+        // register the operator
+        _registerOperatorWithCoordinator(msg.sender, quorumNumbers, pubkey);
```

Comment:
is there a good reason for doing this first? seems more appropriate to do later / at the end, after kicking other(s).

Reply:
instead of calculating the weight of the operator twice for each quorum, we do it once and we look it up

Reply:
mm OK I guess that makes sense. this has some risk of reverting late, but there's a positive gas tradeoff when it doesn't revert.

Reply:
ahh good point, OK 👍 
I think it's reasonable.